build.yaml

```diff
@@ -4913,7 +4926,6 @@ configs:
     DEFINES: NDEBUG
   dbg:
     CPPFLAGS: -O0
-    CXXFLAGS: -fno-exceptions
```

Comment:
is `CatchingFunctionHandler` enough to guarantee we won't accidentally introduce exceptions in the code?

Reply:
No, it's not. It's only one small piece of exception safety. This is specifically to make sure that exceptions in user code called by the sync server don't crash the server. This is the one code path by which we're calling user code.

That said, it relates to the line above only in that we can't even use exception-related features with that command-line flag in place. The code works fine either way, so do you think that the default should be to keep -fno-exceptions and then tell users that if they want to have exceptions they need to remove that line? Or perhaps make a different build target that allows exceptions?

Reply:
I was thinking about users using an already compiled version of gRPC. In that case, the user should be ok and able to use exceptions in her application code regardless of whether gRPC has been compiled with `-fno-exceptions`. Now, if the user is compiling everything -incluiding gRPC-
 from source, then she'd have to create a separate target (or remove that line, given that she's willing to get dirty compiling everything anyway). In general I think erring on the side of caution is best, that is, leaving `-fno-exceptions` there.

On this vein, but going the opposite direction, [this post](https://blog.mozilla.org/nnethercote/2011/01/18/the-dangers-of-fno-exceptions/) presents the dangers of disallowing exceptions on code that may use the std lib. I'm sure you were aware of that already, but just in case.

Reply:
So the issue is that this was exactly what has been requested in #13706 (and before that #9795). Users are writing method handlers for their sync servers, and those method handlers call other things. Our original pushback was that if users call functions that throw exceptions, then they should be responsible for catching any possible exceptions that are thrown (and gRPC can keep on building with `-fno-exceptions`). However, that is an impractical pushback because it violates POLA: something called from gRPC (the user's method handler) can end up crashing the gRPC server. Users may or may not even know if the code they are calling has exceptions if the exception is thrown 20 levels deep in the stack.

Believe me, I am not a fan of exception-based programming. `exception_test` in this PR is the first time I have ever `throw`n in 20+ years of writing C++ code. But it's more important to keep the server alive even if it means sacrificing some purity. Note that `googletest` has [related behavior](https://github.com/google/googletest/blob/7b6561c56e353100aca8458d7bc49c4e0119bae8/googletest/docs/AdvancedGuide.md#disabling-catching-test-thrown-exceptions): the default behavior is that if a test throws an exception, `googletest` catches the exception and marks the test failed but doesn't kill the executable.

Reply:
Because this is essentially a permanent promise, I propose that we create a gRFC and discuss it further before merging this. I propose a target merge date of January 15 so that the gRFC can settle out after a quorum of folks get back from the holidays.